### PR TITLE
Fix negative Kafka partition bug

### PR DESCRIPTION
**Pull request by cbrown184**
## What does this PR do?

Fix a bug in Kafka output "partition to hash method" which results in a negative Kafka partition.
Demo of the bug available on my other branch. https://github.com/cbrown184/beats/tree/negative_partition_edge_case_demo_test

You can reproduce this with any key that gives the hash value of 2147483648. E.g the string "16000002EFEBA11E"

## Why is it important?

When the partition key hash is equal to 2147483648 Filebeat gets stuck in an endless loop.

## Checklist

- [x] My code follows the style guidelines of this project
- [ ] ~~I have commented my code, particularly in hard-to-understand areas~~
- [ ] ~~I have made corresponding changes to the documentation~~
- [ ] ~~I have made corresponding change to the default configuration files~~
- [x] I have added tests that prove my fix is effective or that my feature works
- [x] I have added an entry in `CHANGELOG.next.asciidoc` or `CHANGELOG-developer.next.asciidoc`.

## Author's Checklist

- [ ]

## Related issues
https://discuss.elastic.co/t/negative-kafka-partition-bug/270046/3

This bug was discussed previously in this thread.
https://discuss.elastic.co/t/filebeat-kafka-output-hash-hash-get-negative-partition-cause-stuck/265153

## Use cases

hash(16000002EFEBA11E) = 2147483648
Int32(2147483648) = -2147483648
-(-2147483648) = -2147483648
-2147483648 % 3 = -2
Kafka doesn't have -2 partition, STUCK.
Kafka doesn't have -2 partition, so the kafka producer becomes stuck forever.

## Logs

Filebeat endless loop

2021-02-23T13:18:40.911+0800	DEBUG	[kafka]	kafka/client.go:169	got event.Meta["partition"] = -2
2021-02-23T13:18:40.911+0800	DEBUG	[kafka]	kafka/client.go:179	got event.Meta["topic"] = hashtest
2021-02-23T13:18:40.911+0800	DEBUG	[kafka]	kafka/client.go:273	finished kafka batch
2021-02-23T13:18:40.911+0800	DEBUG	[kafka]	kafka/client.go:287	Kafka publish failed with: kafka: partitioner returned an invalid partition index
2021-02-23T13:18:40.911+0800	DEBUG	[kafka]	kafka/client.go:169	got event.Me

Error Logs in production
2021-04-09T13:48:02Z DBG  Kafka publish failed with: kafka: partitioner returned an invalid partition index

Demo test failing
=== RUN   TestHashMaxIntPlusOneDoesNotReturnNegativePartition
    partition_test.go:328: 
        	Error Trace:	partition_test.go:328
        	Error:      	Should be true
        	Test:       	TestHashMaxIntPlusOneDoesNotReturnNegativePartition
--- FAIL: TestHashMaxIntPlusOneDoesNotReturnNegativePartition (0.00s)

FAIL

Expected :8
Actual   :-8

### CODE REVIEW

File: libbeat/outputs/kafka/partition.go [hunk 1]
```diff
@@ -272,11 +272,11 @@ func makeFieldsHashPartitioner(log *logp.Logger, fields []string, dropFail bool)
 }
 
 func hash2Partition(hash uint32, numPartitions int32) (int32, error) {
-	p := int32(hash)
+	p := int32(hash) % numPartitions
```
Comment (ddcprg) [inline libbeat/outputs/kafka/partition.go:275]: how about 
```
p := hash % numPartitions
return p, nil
```
?

Comment (cbrown184) [reply libbeat/outputs/kafka/partition.go:275]: gives compiler error

./partition.go:275:12: invalid operation: hash % numPartitions (mismatched types uint32 and int32)

**Review by urso — APPROVED** [review]

**Comment by urso** [comment]
Change LGTM. Thank you for fixing it right away!

**Pull request merged**